src/Services/QueuedJobService.php

```diff
@@ -672,28 +703,59 @@ protected function initialiseJob(QueuedJobDescriptor $jobDescriptor)
      */
     protected function grabMutex(QueuedJobDescriptor $jobDescriptor)
     {
-        // write the status and determine if any rows were affected, for protection against a
-        // potential race condition where two or more processes init the same job at once.
-        // This deliberately does not use write() as that would always update LastEdited
-        // and thus the row would always be affected.
         try {
-            DB::query(sprintf(
-                'UPDATE "QueuedJobDescriptor" SET "JobStatus" = \'%s\' WHERE "ID" = %s'
-                . ' AND "JobFinished" IS NULL AND "JobStatus" NOT IN (%s)',
-                QueuedJob::STATUS_INIT,
-                $jobDescriptor->ID,
-                "'" . QueuedJob::STATUS_RUN . "', '" . QueuedJob::STATUS_COMPLETE . "', '"
-                . QueuedJob::STATUS_PAUSED . "', '" . QueuedJob::STATUS_CANCELLED . "'"
-            ));
+            // Start a transaction which will hold until we have a lock on this descriptor.
+            DB::get_conn()->withTransaction(function () use ($jobDescriptor) {
+                $query = 'SELECT `ID`, UUID() as `mutex`, `Worker` FROM `QueuedJobDescriptor`'
+                    . ' WHERE `ID` = %s AND `Worker` IS NULL FOR UPDATE';
+
+                $row = DB::query(sprintf($query, $jobDescriptor->ID))->first();
+
+                // If the row is not returned, then it was reserved by another thread instead.
+                if (!isset($row['mutex'])) {
+                    throw new Exception('Queued Jobs - Obtaining mutex failed (job locked to another thread)');
+                }
+
+                $mutex = $row['mutex'];
+                $workerExpiry = $this->getWorkerExpiry();
+
+                // Otherwise reserve this row.
+                $query = 'UPDATE `QueuedJobDescriptor` SET `Expiry` = \'%s\' , `Worker` = \'%s\','
+                    . ' `WorkerCount` = `WorkerCount` + 1 WHERE `ID` = %s';
+
+                DB::query(sprintf($query, $workerExpiry, $mutex, $jobDescriptor->ID));
+
+                // Make sure we've got our row, and everything is fine.
+                $row = DB::query(
+                    sprintf('SELECT `ID`, `Worker` FROM `QueuedJobDescriptor` WHERE `ID` = %s', $jobDescriptor->ID)
```

Comment:
Same here, bindings and ansi sql please (or better the ORM sql generator to solve both)